rasterio/plot.py

```diff
@@ -33,38 +33,69 @@
 logger = logging.getLogger(__name__)
 
 
-def show(source, cmap='gray', with_bounds=True):
+def show(source, cmap='gray', with_bounds=True, ax=None, title=None, **kwargs):
     """Display a raster or raster band using matplotlib.
 
     Parameters
     ----------
-    source : array-like or (raster dataset, bidx)
+    source : array-like, or (raster dataset, bidx) tuple, or raster dataset, 
         If array-like, should be of format compatible with
         matplotlib.pyplot.imshow. If the tuple (raster dataset, bidx),
-        selects band `bidx` from raster.
+        selects band `bidx` from raster.  If raster dataset display the rgb image 
+        as defined in the colorinterp metadata, or default to first band. 
     cmap : str (opt)
         Specifies the colormap to use in plotting. See
         matplotlib.Colors.Colormap. Default is 'gray'.
     with_bounds : bool (opt)
         Whether to change the image extent to the spatial bounds of the image,
         rather than pixel coordinates. Only works when source is
-        (raster dataset, bidx).
+        (raster dataset, bidx) or raster dataset.
+    ax : matplotlib axes (opt)
+        The raster will be added to this axes if passed.
+    title : str, optional
+        Title for the figure.
+    **kwargs : optional keyword arguments
+        These will be passed to the matplotlib imshow method. See full list at:
+        http://matplotlib.org/api/axes_api.html?highlight=imshow#matplotlib.axes.Axes.imshow
     """
+    if plt is None:  # pragma: no cover
+        raise ImportError("Could not import matplotlib")
+
     if isinstance(source, tuple):
         arr = source[0].read(source[1])
         if with_bounds:
-            extent = plotting_extent(source[0])
+            kwargs['extent'] = plotting_extent(source[0])
+    elif  isinstance(source, rasterio._io.RasterReader):
```

Comment:
If we're going to instance check, we should import `RasterReader` first. Otherwise
```
>>> import rasterio
>>> a = None
>>> isinstance(a, rasterio._io.RasterReader)
Traceback (most recent call last):
  File "<ipython-input-5-c87dc785f86f>", line 1, in <module>
    isinstance(a, rasterio._io.RasterReader)
AttributeError: module 'rasterio' has no attribute '_io'
```

Instead, we should `from rasterio._io import RasterReader` at the top and use it here. @sgillies  and @brendan-ward are working on making a public `rasterio.io` module so it will eventually have to switch to that.

The bigger question is if we **should even do instance checking?** While we don't have a use case for it at the moment, we could potentially have other classes (either in our `io` module or third-part software like xarray) which implement the same interface. Maybe we just check that `source` has the methods we expect, e.g.

```
elif hasattr(source, 'read') and hasattr(source, 'indexes') and hasattr(source, 'colorinterp'):
```